# Multicollinearity in Linear Regression

In [2]:
import pandas as pd

In [9]:

import statsmodels.api as sm #This library consists of Different Regression Model
df_adv = pd.read_csv('Advertising.csv', index_col=0)
X = df_adv[['TV', 'Radio', 'Newspaper']]
y = df_adv['Sales']
df_adv.head()

# sales = b0*1 + b1*tv + b2*radio + b3*newspaper

TV  Radio  Newspaper  Sales
1  230.1   37.8       69.2   22.1
2   44.5   39.3       45.1   10.4
3   17.2   45.9       69.3    9.3
4  151.5   41.3       58.5   18.5
5  180.8   10.8       58.4   12.9

In [10]:
X = sm.add_constant(X)

In [11]:
X

const     TV  Radio  Newspaper
1      1.0  230.1   37.8       69.2
2      1.0   44.5   39.3       45.1
3      1.0   17.2   45.9       69.3
4      1.0  151.5   41.3       58.5
5      1.0  180.8   10.8       58.4
..     ...    ...    ...        ...
196    1.0   38.2    3.7       13.8
197    1.0   94.2    4.9        8.1
198    1.0  177.0    9.3        6.4
199    1.0  283.6   42.0       66.2
200    1.0  232.1    8.6        8.7

[200 rows x 4 columns]

In [12]:
#fit a OLS model with intercept on TV and Radio

model = sm.OLS(y, X).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Sales   R-squared:                       0.897
Model:                            OLS   Adj. R-squared:                  0.896
Method:                 Least Squares   F-statistic:                     570.3
Date:                Wed, 16 Jun 2021   Prob (F-statistic):           1.58e-96
Time:                        10:57:29   Log-Likelihood:                -386.18
No. Observations:                 200   AIC:                             780.4
Df Residuals:                     196   BIC:                             793.6
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          2.9389      0.312      9.422      0.000       2.324       3.554
TV             0.0458      0.001     32.809      0.000       0.043       0.049
Radio          0.1885      0.009     21.893      0.000       0.172       0.206
Newspaper     -0.0010      0.006     -0.177      0.860      -0.013       0.011
==============================================================================
Omnibus:                       60.414   Durbin-Watson:                   2.084
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              151.241
Skew:                          -1.327   Prob(JB):                     1.44e-33
Kurtosis:                       6.332   Cond. No.                         454.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [16]:
import matplotlib.pyplot as plt
X.iloc[:,1:].corr()
#.corr() is used to evaluate correlation between the variables

TV     Radio  Newspaper
TV         1.000000  0.054809   0.056648
Radio      0.054809  1.000000   0.354104
Newspaper  0.056648  0.354104   1.000000

In [18]:
df_salary = pd.read_csv('Salary_Data.csv')
df_salary.head()

YearsExperience   Age  Salary
0              1.1  21.0   39343
1              1.3  21.5   46205
2              1.5  21.7   37731
3              2.0  22.0   43525
4              2.2  22.2   39891

In [23]:
X = df_salary[['YearsExperience', 'Age']]
y = df_salary['Salary']
# salary =  b0*1 + b1*YearsExperienece + b2*Age

In [25]:
X = sm.add_constant(X)
X

const  YearsExperience   Age
0     1.0              1.1  21.0
1     1.0              1.3  21.5
2     1.0              1.5  21.7
3     1.0              2.0  22.0
4     1.0              2.2  22.2
5     1.0              2.9  23.0
6     1.0              3.0  23.0
7     1.0              3.2  23.3
8     1.0              3.2  23.3
9     1.0              3.7  23.6
10    1.0              3.9  23.9
11    1.0              4.0  24.0
12    1.0              4.0  24.0
13    1.0              4.1  24.0
14    1.0              4.5  25.0
15    1.0              4.9  25.0
16    1.0              5.1  26.0
17    1.0              5.3  27.0
18    1.0              5.9  28.0
19    1.0              6.0  29.0
20    1.0              6.8  30.0
21    1.0              7.1  30.0
22    1.0              7.9  31.0
23    1.0              8.2  32.0
24    1.0              8.7  33.0
25    1.0              9.0  34.0
26    1.0              9.5  35.0
27    1.0              9.6  36.0
28    1.0             10.3  37.0
29    1.0             10.5  38.0

In [27]:
model = sm.OLS(y, X).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 Salary   R-squared:                       0.960
Model:                            OLS   Adj. R-squared:                  0.957
Method:                 Least Squares   F-statistic:                     323.9
Date:                Wed, 16 Jun 2021   Prob (F-statistic):           1.35e-19
Time:                        11:26:36   Log-Likelihood:                -300.35
No. Observations:                  30   AIC:                             606.7
Df Residuals:                      27   BIC:                             610.9
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
===================================================================================
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
const           -6661.9872   2.28e+04     -0.292      0.773   -5.35e+04    4.02e+04
YearsExperience  6153.3533   2337.092      2.633      0.014    1358.037    1.09e+04
Age              1836.0136   1285.034      1.429      0.165    -800.659    4472.686
==============================================================================
Omnibus:                        2.695   Durbin-Watson:                   1.711
Prob(Omnibus):                  0.260   Jarque-Bera (JB):                1.975
Skew:                           0.456   Prob(JB):                        0.372
Kurtosis:                       2.135   Cond. No.                         626.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [28]:
import matplotlib.pyplot as plt
X.iloc[:,1:].corr()

YearsExperience       Age
YearsExperience         1.000000  0.987258
Age                     0.987258  1.000000

Since both are highly correlated now we have to emit one which has higher p value. Here Age = 0.165 has high p-value so it will be omitted. 

In [30]:
X = df_salary[['YearsExperience']]
X = sm.add_constant(X)
X

const  YearsExperience
0     1.0              1.1
1     1.0              1.3
2     1.0              1.5
3     1.0              2.0
4     1.0              2.2
5     1.0              2.9
6     1.0              3.0
7     1.0              3.2
8     1.0              3.2
9     1.0              3.7
10    1.0              3.9
11    1.0              4.0
12    1.0              4.0
13    1.0              4.1
14    1.0              4.5
15    1.0              4.9
16    1.0              5.1
17    1.0              5.3
18    1.0              5.9
19    1.0              6.0
20    1.0              6.8
21    1.0              7.1
22    1.0              7.9
23    1.0              8.2
24    1.0              8.7
25    1.0              9.0
26    1.0              9.5
27    1.0              9.6
28    1.0             10.3
29    1.0             10.5

In [32]:
model = sm.OLS(y, X).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 Salary   R-squared:                       0.957
Model:                            OLS   Adj. R-squared:                  0.955
Method:                 Least Squares   F-statistic:                     622.5
Date:                Wed, 16 Jun 2021   Prob (F-statistic):           1.14e-20
Time:                        11:31:52   Log-Likelihood:                -301.44
No. Observations:                  30   AIC:                             606.9
Df Residuals:                      28   BIC:                             609.7
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
===================================================================================
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
const            2.579e+04   2273.053     11.347      0.000    2.11e+04    3.04e+04
YearsExperience  9449.9623    378.755     24.950      0.000    8674.119    1.02e+04
==============================================================================
Omnibus:                        2.140   Durbin-Watson:                   1.648
Prob(Omnibus):                  0.343   Jarque-Bera (JB):                1.569
Skew:                           0.363   Prob(JB):                        0.456
Kurtosis:                       2.147   Cond. No.                         13.2
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""